common/src/main/java/org/apache/drill/common/types/Types.java

```diff
@@ -636,43 +658,63 @@ public static String toString(final MajorType type) {
 
   /**
    * Get the <code>precision</code> of given type.
-   * @param majorType
-   * @return
+   *
+   * @param majorType major type
+   * @return precision value
    */
   public static int getPrecision(MajorType majorType) {
-    MinorType type = majorType.getMinorType();
-
-    if (type == MinorType.VARBINARY || type == MinorType.VARCHAR) {
-      return 65536;
-    }
-
     if (majorType.hasPrecision()) {
       return majorType.getPrecision();
     }
 
-    return 0;
+    return isScalarStringType(majorType) ? MAX_VARCHAR_LENGTH : UNDEFINED;
```

Comment:
I believe what this PR tries to do is exactly as what you describe : derive length based on metadata. Such metadata comes from the function used , or the string literal.  For example, if you have a cast(some_input as varchar(20)), you will know the output type's max length is 20 (depending on the input, the output may only have values up to 15 characters).   

The max length is different from what could be obtained from {{RecordBatchSizer}}.